File: 1/2020.09.23/206.Reverse Linked List/python/solution.py
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # recursively
        if not head or not head.next:
            return head
        p = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return p

    def reverseList(self, head: ListNode) -> ListNode:
        # iteratively
        dummy = None
        while head:
            p = head.next
            head.next, dummy = dummy, head
            head = p
        return dummy

File: 1/2020.09.23/206.Reverse Linked List/java/solution.java
class Solution {
    public ListNode reverseList(ListNode head) {
        // iteratively
        ListNode dummy = null;
        while (head != null) {
            ListNode p = head.next;
            head.next = dummy;
            dummy = head;
            head = p;
        }
        return dummy;
    }

    public ListNode reverseList(ListNode head) {
        // recursively
        if (head == null || head.next == null) {
            return head;
        }
        ListNode p = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return p;
    }
}
